stdlib/public/core/UnicodeHelpers.swift

```diff
@@ -122,6 +122,46 @@ internal func _utf8ScalarLength(_ x: UInt8) -> Int {
   return (~x).leadingZeroBitCount
 }
 
+@inlinable @inline(__always)
+internal func _utf16Length(_ rawBuffer: UnsafeRawBufferPointer) -> Int {
+  guard rawBuffer.count > 0 else { return 0 }
+  _internalInvariant(!UTF8.isContinuation(rawBuffer.first.unsafelyUnwrapped))
+  _internalInvariant(
+       UTF8.isASCII(rawBuffer.last.unsafelyUnwrapped)
+    || UTF8.isContinuation(rawBuffer.last.unsafelyUnwrapped))
+  let wordASCIIMask = UInt(truncatingIfNeeded: 0x8080_8080_8080_8080 as UInt64)
+  var utf16Count = 0
+  var readIdx = 0
+
+  while readIdx + MemoryLayout<UInt>.stride < rawBuffer.count {
+    let maybeASCII = Builtin.loadRaw((rawBuffer.baseAddress._unsafelyUnwrappedUnchecked + readIdx)._rawValue) as UInt
```

Comment:
Is this how to spell an unaligned load? I think we should rather put this in a utility function that is named to indicate this.

(Would arranging for aligned loads be better? IIRC, Hasher's buffer combining method aligns its loads, but I don't recall if I ever benchmarked it.)

I wonder if there is some neat simd-style trick to count bytes that don't have `10` in their most significant bits. E.g., `((word & 0xC0C0C0C0...) ^ 0x80808080...)` reduces it to counting non-zero bytes. Hm, this could be promising... 🤔

Reply:
Unaligned loads are going through S-E shortly :D I'm just gonna wait for that. And yeah I would love to find some good tricks here.

I didn't do aligned loads because it's easy to fall back into unaligned if the non-ascii cleanup loop ends in a multibyte character, but it would be possible to fix that up if it turned out to be profitable.